temporal-sdk/src/main/java/io/temporal/internal/worker/TrackingSlotSupplier.java

```diff
@@ -0,0 +1,219 @@
+ */
+
+package io.temporal.internal.worker;
+
+import com.uber.m3.tally.NoopScope;
+import com.uber.m3.tally.Scope;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.tuning.*;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Wraps a slot supplier and supplements it with additional tracking information that is useful to
+ * provide to all implementations. This type is used internally rather than {@link SlotSupplier}
+ * directly.
+ *
+ * @param <SI> The slot info type
+ */
+public class TrackingSlotSupplier<SI extends SlotInfo> {
+  private final SlotSupplier<SI> inner;
+  private final AtomicInteger issuedSlots = new AtomicInteger();
+  private final Map<SlotPermit, SI> usedSlots = new ConcurrentHashMap<>();
+  private Scope metricsScope;
+
+  public TrackingSlotSupplier(SlotSupplier<SI> inner) {
+    this.inner = inner;
+    metricsScope = new NoopScope();
+  }
+
+  public SlotPermit reserveSlot(SlotReservationData dat) throws InterruptedException {
+    SlotPermit p = inner.reserveSlot(createCtx(dat));
+    issuedSlots.incrementAndGet();
+    return p;
+  }
+
+  public Optional<SlotPermit> tryReserveSlot(SlotReservationData dat) {
+    Optional<SlotPermit> p = inner.tryReserveSlot(createCtx(dat));
+    if (p.isPresent()) {
+      issuedSlots.incrementAndGet();
+    }
+    return p;
+  }
+
+  public void markSlotUsed(SI slotInfo, SlotPermit permit) {
+    if (permit == null) {
+      return;
+    }
+    inner.markSlotUsed(new SlotMarkUsedContextImpl(slotInfo, permit));
+    usedSlots.put(permit, slotInfo);
+    publishSlotsMetric();
+  }
+
+  public void releaseSlot(SlotReleaseReason reason, SlotPermit permit) {
+    if (permit == null) {
+      return;
+    }
+    SI slotInfo = usedSlots.get(permit);
+    inner.releaseSlot(new SlotReleaseContextImpl(reason, permit, slotInfo));
+    issuedSlots.decrementAndGet();
+    usedSlots.remove(permit);
+    publishSlotsMetric();
+  }
+
+  public int maximumSlots() {
+    return inner.getMaximumSlots();
+  }
+
+  public int getIssuedSlots() {
+    return issuedSlots.get();
+  }
+
+  public void setMetricsScope(Scope metricsScope) {
+    this.metricsScope = metricsScope;
+    publishSlotsMetric();
+  }
+
+  /**
+   * If any slot supplier is resource-based, we want to attach a metrics scope to the controller
+   * (before it's labelled with the worker type).
+   */
+  public void attachMetricsToResourceController(Scope metricsScope) {
+    if (inner instanceof ResourceBasedSlotSupplier) {
+      ((ResourceBasedSlotSupplier<?>) inner).getResourceController().setMetricsScope(metricsScope);
+    }
+  }
+
+  Map<SlotPermit, SI> getUsedSlots() {
+    return usedSlots;
+  }
+
+  private void publishSlotsMetric() {
+    this.metricsScope
+        .gauge(MetricsType.WORKER_TASK_SLOTS_AVAILABLE)
+        .update(maximumSlots() - usedSlots.size());
```

Comment:
If `inner` does not implement `maximumSlots()` then I don't think we should be emitting this metric. May need to modify the definition of `maximumSlots` to help indicate if the underlying slot supplier actually has a fixed size

Reply:
Yeah, I thought about this and can't remember why I decided it was still worth doing, but I'll make that change.